---
title: "logoplots_by_residue"
output: html_document
date: "2024-07"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##Load librarys
```{r}
library(Biostrings) 
library(tidyverse)
library(motifStack)
library(DiffLogo)
```


# set up a custom color scheme

this color scheme mimics the "chemistry" color scheme from motifStack (defined [here](https://github.com/jianhong/motifStack/blob/5aa80388b44bc8f93738315310cfb56c3495130c/R/publicUtilities.R))

```{r}
## some quick code I used when I was looking at the motifStack color schemes, to figure out what the hex type colors (e.g. `#800080`) actually look like
## plot using hex color
# plot(1:10,1:10, pch=19, cex=2, col="#800080")
## check what I think it is
# points(1:10,0:9, pch=19, cex=3, col="magenta4")
````

```{r set up color scheme}
# changeColors is a function I wrote that works on objects of class "Alphabet" to change the colors for some amino acids (or nucleotides)
# myAlphabet is the Alphabet object (the color scheme we're working on)
# myAA is a vector of the AA (or nuc) we want to change
# myCol is the new color (e.g. "black")
changeColors <- function(myAlphabet, myAA, myCol) {
    # get indices of things we want to change
    whichChar <- which(myAlphabet$chars %in% myAA)
    # get existing color scheme
    tempColors <- myAlphabet$cols
    # change the ones we selected
    tempColors[whichChar] <- myCol
    # add the altered color scheme back to the Alphabet object
    myAlphabet$cols <- tempColors
    # done
    return(myAlphabet)
}

## reportColors is a small function to more easily visualize the color scheme for an Alphabet object 
reportColors <- function(myAlphabet) {
    tibble(aa=myAlphabet$chars, 
           color=myAlphabet$cols) %>% 
        arrange(color)
}
```


```{r}
### ASN is a DiffLogo object (see ?ASN) that defines their default color scheme
### make custom version of the ASN Alphabet object called FULL_ALPHABET_JYchemistryColors
ASN_JYchemistryColors <- ASN

ASN_JYchemistryColors <- changeColors(ASN_JYchemistryColors,
                                      c("A","F","I","L","M","P","V","W"), 
                                      "black")

ASN_JYchemistryColors <- changeColors(ASN_JYchemistryColors, 
                                      c("C","G","S","T","Y"), 
                                      "forest green")

ASN_JYchemistryColors <- changeColors(ASN_JYchemistryColors, 
                                      c("D","E"), 
                                      "red3")

ASN_JYchemistryColors <- changeColors(ASN_JYchemistryColors, 
                                      c("H","K","R"), 
                                      "blue3")

ASN_JYchemistryColors <- changeColors(ASN_JYchemistryColors, 
                                      c("N","Q"), 
                                      "magenta4")

## check it looks how I think it should
reportColors(ASN_JYchemistryColors)
```

##Load data
```{r}
## xx moved this:
# ", show_col_types = FALSE" 
# before the parentheses, so that it goes with read_csv
h5n1screen <- as.data.frame(read_csv('/Users/rgeiger/Dropbox/Data/R_files/240718_vn04screen_recompiled.csv', show_col_types = FALSE))
```

##Define alphabet
```{r}
myLetters <- c(AA_ALPHABET[1:20])
```

##Calculate mean fold change to EV
```{r}
h5n1screen <- as.tibble(h5n1screen) %>% 
    mutate( ff_ren_norm = ff/ren) %>% 
    relocate(ff_ren_norm, .after = ren)

#create categories that can be used for filling bar plot
h5n1screen <- h5n1screen %>% 
    mutate(
        category = case_when(
            var_id == 'ev' ~ 'empty_vector',
            var_id == 'wt' ~ 'wtMxA',
            var_id == 't103a' ~ 'cat_dead',
            var_id == 'nls' ~ 'nls',
            var_seq == 'frameshift' ~ 'frameshift_mutation',
            var_seq == 'double' ~ 'double_colony',
            var_seq == 'stop' ~ 'early_stop_codon',
            .default = 'mxa_variant')
    )


#using ev category to calculate average ev 
empty_vector_data <- h5n1screen %>% 
    filter(category == 'empty_vector')

empty_vector_aves <- empty_vector_data %>% 
    group_by(batch) %>% 
    summarise( empty_vector_mean = mean(ff_ren_norm))

h5n1screen <- left_join( h5n1screen, 
                         empty_vector_aves,
                         by="batch") %>% 
    relocate( empty_vector_mean, .after = ff_ren_norm)

h5n1screen <- h5n1screen %>% 
    mutate(norm_to_EV = empty_vector_mean/ff_ren_norm) %>% 
    relocate(norm_to_EV, .after = empty_vector_mean)
```

##Calculate mean fold restriction per variant_id per batch
```{r}
average_ev_norms <- h5n1screen %>% 
    group_by(batch, var_id) %>% 
    summarise(mean_norm_to_EV = mean(norm_to_EV))

h5n1screen <- left_join( h5n1screen, 
                         average_ev_norms,
                         join_by('batch', 'var_id')) %>% 
    relocate(mean_norm_to_EV, .after = norm_to_EV)
```

##Get rid of individual values so all we have are variants and their mean restriction
```{r}
var_average_restriction <- h5n1screen %>% 
    select(var_id, var_seq, mean_norm_to_EV, category) %>% 
    distinct()

```

##Split sequences into amino acid positions
```{r}
var_only_restriction_sequence <- var_average_restriction %>% 
    filter(category == 'mxa_variant') %>% 
    separate(var_seq, c('weird_blank', 'aa540', 'aa561', 'aa564', 'aa566', 'aa567'), sep = "", remove = FALSE, extra = 'merge', fill = 'right') %>% 
    filter(aa540 != '_') %>% 
    mutate_at(vars(aa540:aa567), toupper)
```

##Create new category for non-restrictors, restrictors, and super-restrictors
```{r}
var_only_restriction_sequence <- var_only_restriction_sequence %>% 
    mutate(restrictor_level=case_when(
        mean_norm_to_EV < 1.5993145 ~ 'non-restrictor',
        mean_norm_to_EV <= 14.8672901 & mean_norm_to_EV >= 1.5993145 ~ 'restrictor',
        mean_norm_to_EV > 14.8672901 ~ 'super-restrictor'))
```


##Position count matrix
```{r}
AAcounts_whole_library <- apply( var_only_restriction_sequence[,4:8], 2, function(x) {
    y <- factor(x, levels=myLetters) 
    table(y)
})

AAcounts_whole_library <- t(AAcounts_whole_library)
```

##Position frequency matrix
```{r}
PFM_whole_library <- pcm2pfm(t(AAcounts_whole_library))

colSums(PFM_whole_library)
```

##Plot function
```{r}
tempPlotFunction <- function(myMatrix, myTitle) {
    tempMotif <- new("pcm", mat=myMatrix, name=myTitle, 
                     color=colorset(alphabet="AA",colorScheme="chemistry"))
    plot(tempMotif, ic.scale=FALSE)
}

#Whole library logo plot
tempPlotFunction(t(AAcounts_whole_library), myTitle="Whole_Library")
```

##Get SRs only -> make PFM -> make DiffLogo to whole_library
```{r, fig.height=11, fig.width=7}
SRs <- var_only_restriction_sequence %>% 
    filter(restrictor_level == 'super-restrictor')

AAcounts_SRs_toEV <- apply( SRs[,4:8], 2, function(x) {
    y <- factor(x, levels=myLetters) 
    table(y, useNA="always")
})

AAcounts_SRs_toEV <- t(AAcounts_SRs_toEV)[,1:20]

PFM_SRs_toEV <- pcm2pfm(t(AAcounts_SRs_toEV))
colSums(PFM_whole_library)

##Below is from ChatGPT
# Run diffLogoFromPwm
## xxx sumProbabilities should work for a seqLogo but I'm not sure it will work for a diffLogo
diffLogoFromPwm(
    pwm1 = PFM_whole_library[ASN$chars, ],
    pwm2 = PFM_SRs_toEV[ASN$chars, ],
    # stackHeight = sumProbabilities(),
    alphabet = ASN)
```

```{r, fig.height=11, fig.width=7}
### make plot to show in markdown doc
diffLogoFromPwm(
    pwm1 = PFM_whole_library[ASN$chars, ],
    pwm2 = PFM_SRs_toEV[ASN$chars, ],
    ymin=0.1, ymax=-0.1,
    alphabet = ASN_JYchemistryColors)

### make exactly the same plot but save to a pdf file
# open the pdf
pdf(file="temp.pdf", height=11, width=7)
# make the plot
diffLogoFromPwm(
    pwm1 = PFM_whole_library[ASN$chars, ],
    pwm2 = PFM_SRs_toEV[ASN$chars, ],
    ymin=0.1, ymax=-0.1,
    alphabet = ASN_JYchemistryColors)
# close the pdf
dev.off()
```

Same logo plot but use custom color scheme
```{r}
fakeHeightFunc <- function(x,y) {
    list(height=2,
         ylab="fake height")
}

diffLogoFromPwm(
    pwm1 = PFM_whole_library[ASN$chars, ],
    pwm2 = PFM_SRs_toEV[ASN$chars, ],
    stackHeight = fakeHeightFunc,
    alphabet = ASN_JYchemistryColors)
```

```{r}
diffLogoFromPwm_custom <- function (
        pwm1, pwm2, 
        ymin = 0, ymax = 0, yaxis_label="Jensen-Shannon divergence",
        stackHeight = shannonDivergence, 
        baseDistribution = normalizedDifferenceOfProbabilities, 
        sparse = FALSE, 
        alphabet = DNA, align_pwms = FALSE, 
        unaligned_penalty = divergencePenaltyForUnaligned, 
        try_reverse_complement = TRUE, base_distribution = NULL, 
        length_normalization = FALSE,
        returnDiffLogoObj=FALSE) {
    diffLogoObj = createDiffLogoObject(
        pwm1, pwm2, stackHeight = stackHeight, 
        baseDistribution = baseDistribution, alphabet = alphabet, 
        align_pwms = align_pwms, unaligned_penalty = unaligned_penalty, 
        try_reverse_complement = try_reverse_complement, base_distribution = NULL, 
        length_normalization = length_normalization)
    if(returnDiffLogoObj) {return(diffLogoObj)}
    diffLogo_custom(diffLogoObj, 
                    ymin = ymin, ymax = ymax, yaxis_label=yaxis_label,
                    xaxis_labels=colnames(pwm1),
                    sparse = sparse)
}

diffLogo_custom <- function (diffLogoObj, 
                             ymin = 0, ymax = 0, yaxis_label="Jensen-Shannon divergence",
                             sparse = FALSE, 
                             xaxis_labels=NULL,
                             diffLogoConfiguration = list()) {
    if (!is(diffLogoObj, "DiffLogo")) {
        msg = paste("Expected DiffLogo, but got ", class(diffLogoObj), 
                    ". Use #createDiffLogoObject to get an DiffLogo from two PWMs.", 
                    sep = "")
        stop(msg)
    }
    ## if we didn't supply ymin/ ymax figure it out
    if (ymin == 0) {
        ymin = diffLogoObj$ylim.posMax
    }
    if (ymax == 0) {
        ymax = diffLogoObj$ylim.negMax
    }
    ylab = diffLogoObj$ylab
    ## before we plot the data, I think this makes blank plot
    if (sparse) {
        plot(NA, xlim = c(0.5, diffLogoObj$npos + 0.5), 
             ylim = c(ymin,ymax), xaxt = "n", 
             ylab = "", mgp = c(0, 0.35, 0), 
             tck = -0.02, cex.axis = 0.8, frame.plot = FALSE, 
             xlab = "")
    }
    else {
        plot(NA, xlim = c(0.5, diffLogoObj$npos + 0.5), 
             ylim = c(ymin, ymax), xaxt = "n", 
             ylab = yaxis_label, 
             xlab = "", ## was "Position"
             frame.plot = FALSE, las=1)
    }
    ## make the x-axis
    if (sparse) {
        axis(1, labels = c(rep("", diffLogoObj$npos)), at = (1:diffLogoObj$npos), 
             tck = -0.02)
        axis(1, labels = c("", ""), 
             at = c(0, (diffLogoObj$npos + 1)), tck = -0)
    }
    else {
        axis(side=1, 
             labels = xaxis_labels, 
             at = (1:diffLogoObj$npos),
             lwd=5)
        ## original x-axis labels
        # axis(side=1, labels = c(1:diffLogoObj$npos), at = (1:diffLogoObj$npos))
        axis(side=1, labels = c("", ""),
             at = c(0, (diffLogoObj$npos + 1)), tck = -0, lwd=5)
    }
    ## make the letters
    polygon(diffLogoObj$letters, col = diffLogoObj$letters$col, 
            border = FALSE)
    if (!is.null(diffLogoObj$unaligned_from_left) && diffLogoObj$unaligned_from_left > 
        0) {
        rect(0.5, -ymin, diffLogoObj$unaligned_from_left + 0.5, 
             -ymax, col = "gray", border = "gray")
    }
    if (!is.null(diffLogoObj$unaligned_from_right) && diffLogoObj$unaligned_from_right > 
        0) {
        rect(diffLogoObj$npos - diffLogoObj$unaligned_from_right + 
                 0.5, -ymin, diffLogoObj$npos + 0.5, -ymax, col = "gray", 
             border = "gray")
    }
    if (!is.null(diffLogoObj$pvals)) {
        leftOffset = 0
        if (!is.null(diffLogoObj$unaligned_from_left)) {
            leftOffset = diffLogoObj$unaligned_from_left
        }
        if (!is.null(diffLogoObj$unaligned_from_right)) {
            rightOffset = diffLogoObj$unaligned_from_right
        }
        npos = ncol(diffLogoObj$pwm1)
        for (j in (leftOffset + 1):(npos - rightOffset)) {
            if (diffLogoObj$pvals[j] < 0.05) {
                text(j, ymin, "*")
            }
        }
    }
    # this makes the y=0 line
    # lwd is line width
    lines(x=c(0, diffLogoObj$npos+0.5), 
          y=c(0, 0),
          lwd=5, col="lightgray")
}
```

```{r}
# 
# temp <- diffLogoFromPwm_custom(
#     pwm1 = PFM_whole_library[ASN$chars, ],
#     pwm2 = PFM_SRs_toEV[ASN$chars, ],
#     ymin=0.1, ymax=-0.1,
#     alphabet = ASN_JYchemistryColors,
#     returnDiffLogoObj=TRUE)


diffLogoFromPwm_custom(
    pwm1 = PFM_whole_library[ASN$chars, ],
    pwm2 = PFM_SRs_toEV[ASN$chars, ],
    ymin=0.1, ymax=-0.1,
    alphabet = ASN_JYchemistryColors)
```

#Make DiffLogos for each bin of 561 mutants
##Filter for 561F only
```{r, fig.height=11, fig.width=7}
AAcounts_561F <- var_only_restriction_sequence %>% 
    filter(aa561 == 'F')

AAcounts_561F_4sites <- AAcounts_561F %>% 
    select(aa540, aa564, aa566, aa567)

AAcounts_561F_4sites <- apply( AAcounts_561F_4sites[,1:4], 2, function(x) {
    y <- factor(x, levels=myLetters) 
    table(y)
})

AAcounts_561F_4sites <- t(AAcounts_561F_4sites)

#Position frequency matrix

PFM_561F_4sites <- pcm2pfm(t(AAcounts_561F_4sites))

SRs_561F <- var_only_restriction_sequence %>% 
    filter(restrictor_level == 'super-restrictor', aa561 == 'F')

AAcounts_561F_SRs_4sites <- SRs_561F %>% 
    select(aa540, aa564, aa566, aa567)

AAcounts_561F_SRs_4sites <- apply( AAcounts_561F_SRs_4sites[,1:4], 2, function(x) {
    y <- factor(x, levels=myLetters) 
    table(y, useNA="always")
})

AAcounts_561F_SRs_4sites <- t(AAcounts_561F_SRs_4sites)[,1:20]

PFM_561F_SRs_4sites <- pcm2pfm(t(AAcounts_561F_SRs_4sites))

colSums(PFM_561F_4sites)
colSums(PFM_561F_SRs_4sites)

##Below is from ChatGPT
# Run diffLogoFromPwm
diffLogoFromPwm(
    pwm1 = PFM_561F_4sites[ASN$chars, ],
    pwm2 = PFM_561F_SRs_4sites[ASN$chars, ],
    #stackHeight = sumProbabilities(1.0),
    alphabet = ASN)
```

##Filter for 561W only
```{r, fig.height=11, fig.width=7}
AAcounts_561W <- var_only_restriction_sequence %>% 
    filter(aa561 == 'W')

AAcounts_561W_4sites <- AAcounts_561W %>% 
    select(aa540, aa564, aa566, aa567)

AAcounts_561W_4sites <- apply( AAcounts_561W_4sites[,1:4], 2, function(x) {
    y <- factor(x, levels=myLetters) 
    table(y)
})

AAcounts_561W_4sites <- t(AAcounts_561W_4sites)

#Position frequency matrix

PFM_561W_4sites <- pcm2pfm(t(AAcounts_561W_4sites))

SRs_561W <- var_only_restriction_sequence %>% 
    filter(restrictor_level == 'super-restrictor', aa561 == 'W')

AAcounts_561W_SRs_4sites <- SRs_561W %>% 
    select(aa540, aa564, aa566, aa567)

AAcounts_561W_SRs_4sites <- apply( AAcounts_561W_SRs_4sites[,1:4], 2, function(x) {
    y <- factor(x, levels=myLetters) 
    table(y, useNA="always")
})

AAcounts_561W_SRs_4sites <- t(AAcounts_561W_SRs_4sites)[,1:20]

PFM_561W_SRs_4sites <- pcm2pfm(t(AAcounts_561W_SRs_4sites))

colSums(PFM_561W_4sites)
colSums(PFM_561W_SRs_4sites)

##Below is from ChatGPT
# Run diffLogoFromPwm
diffLogoFromPwm(
    pwm1 = PFM_561W_4sites[ASN$chars, ],
    pwm2 = PFM_561W_SRs_4sites[ASN$chars, ],
    ymin=0.1, ymax=-0.1,
    #stackHeight = sumProbabilities(1.0),
    alphabet = ASN)
```

##Filter for 561Y only
```{r, fig.height=11, fig.width=7}
AAcounts_561Y <- var_only_restriction_sequence %>% 
    filter(aa561 == 'Y')

AAcounts_561Y_4sites <- AAcounts_561Y %>% 
    select(aa540, aa564, aa566, aa567)

AAcounts_561Y_4sites <- apply( AAcounts_561Y_4sites[,1:4], 2, function(x) {
    y <- factor(x, levels=myLetters) 
    table(y)
})

AAcounts_561Y_4sites <- t(AAcounts_561Y_4sites)

#Position frequency matrix

PFM_561Y_4sites <- pcm2pfm(t(AAcounts_561Y_4sites))

SRs_561Y <- var_only_restriction_sequence %>% 
    filter(restrictor_level == 'super-restrictor', aa561 == 'Y')

AAcounts_561Y_SRs_4sites <- SRs_561Y %>% 
    select(aa540, aa564, aa566, aa567)

AAcounts_561Y_SRs_4sites <- apply( AAcounts_561Y_SRs_4sites[,1:4], 2, function(x) {
    y <- factor(x, levels=myLetters) 
    table(y, useNA="always")
})

AAcounts_561Y_SRs_4sites <- t(AAcounts_561Y_SRs_4sites)[,1:20]

PFM_561Y_SRs_4sites <- pcm2pfm(t(AAcounts_561Y_SRs_4sites))

colSums(PFM_561Y_4sites)
colSums(PFM_561Y_SRs_4sites)

##Below is from ChatGPT
# Run diffLogoFromPwm
diffLogoFromPwm(
    pwm1 = PFM_561Y_4sites[ASN$chars, ],
    pwm2 = PFM_561Y_SRs_4sites[ASN$chars, ],
    #stackHeight = sumProbabilities(1.0),
    alphabet = ASN)
```

##How many variants total have 561F/W/Y versus how many super-restrictors have F/W/Y?
```{r}
#Create a data frame with columns F/W/Y and rows whole_library and super-restrictors
FWY_distribution <- as.data.frame(t(AAcounts_whole_library))  %>% 
    filter(aa561 != 0) %>% 
    select(aa561) %>% 
    rownames_to_column('aa561_seq')

FWY_SR_distribution <- as.data.frame(t(AAcounts_SRs_toEV))  %>% 
    filter(aa561 != 0) %>% 
    select(aa561) %>% 
    rownames_to_column('aa561_seq')

FWY_distribution_comparison <- full_join(
    FWY_distribution, FWY_SR_distribution,
    join_by('aa561_seq'),
    suffix = c('_whole_library', '_SRs'))

t(FWY_distribution_comparison)
```

